modules/data.atmosphere/R/upscale_met.R

```diff
@@ -0,0 +1,107 @@
+##' upscale_met upscales the temporal resolution of a dataset
+##' @name upscale_met
+##' @title upscale_met
+##' @export
+##' @param outfolder
+##' @param input_met - met dataset
+##' @param site.id
+##' @param reso - numeric value in HOURS that is the desired resolution you wish the dataset to be upscaled to
+##' @param overwrite
+##' @param verbose
+##' @author James Simkins
+
+upscale_met <- function(outfolder, input_met, site.id, reso = 6, overwrite = FALSE, 
+                        verbose = FALSE, ...) {
+  
+  substrRight <- function(x, n) {
+    substr(x, nchar(x) - n + 1, nchar(x))
+  }
+  sub_str <- substrRight(input_met, 7)
+  year <- substr(sub_str, 1, 4)
+  year <- as.numeric(year)
+  
+  # Variable names
+  var <- data.frame(CF.name <- c("air_temperature", "air_temperature_max", "air_temperature_min", 
+                                 "surface_downwelling_longwave_flux_in_air", "air_pressure", "surface_downwelling_shortwave_flux_in_air", 
+                                 "eastward_wind", "northward_wind", "specific_humidity", "precipitation_flux"), 
+                    units <- c("Kelvin", "Kelvin", "Kelvin", "W/m2", "Pascal", "W/m2", "m/s", 
+                               "m/s", "g/g", "kg/m2/s"))
+  # Reading in the data
+  data <- list()
+  tem <- ncdf4::nc_open(input_met)
+  dim <- tem$dim
+  for (j in seq_along(var$CF.name)) {
+    if (exists(as.character(var$CF.name[j]), tem$var) == FALSE) {
+      data[[j]] <- NA
+    } else {
+      data[[j]] <- ncdf4::ncvar_get(tem, as.character(var$CF.name[j]))
+    }
+  }
+  
+  lat_data <- as.numeric(ncdf4::ncvar_get(tem, "latitude"))
+  lon_data <- as.numeric(ncdf4::ncvar_get(tem, "longitude"))
+  ncdf4::nc_close(tem)
+  
+  data <- data.frame(data)
+  colnames(data) <- var$CF.name
+  
+  if (lubridate::leap_year(year) == TRUE) {
+    sp <- 366
+  }
+  if (lubridate::leap_year(year) == FALSE) {
+    sp <- 365
+  }
+  reso_len <- sp * 24/reso
+  
+  step <- length(data$air_temperature)/reso_len
+  upscale_data <- data.frame()
+  for (n in 1:length(var$CF.name)) {
+    for (x in 1:reso_len) {
+      upscale_data[x, n] <- mean(data[(x * step - step + 1):(x * step), n])
+    }
+  }
```

Comment:
Rather than two loops, it would be much more efficient to do this with apply (over variables) and tapply (over an indicator variable). e.g. ```apply(data,2,function(x){tapply(x,indicator,mean)})```